istioctl/pkg/auth/upgrader.go

```diff
@@ -0,0 +1,376 @@
+	}
+	sources := convertBindingToSources(v1alpha1Model.Principals)
+	// If there is no services field or it's defined as "*", it's equivalent to an AuthorizationPolicy
+	// with no workload selector
+	authzPolicyConfig := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:      schemas.AuthorizationPolicy.Type,
+			Namespace: namespace,
+		},
+		Spec: &rbac_v1beta1.AuthorizationPolicy{
+			Selector: &v1beta1.WorkloadSelector{},
+			Rules: []*rbac_v1beta1.Rule{
+				{
+					From: sources,
+					To: []*rbac_v1beta1.Rule_To{
+						{
+							Operation: operations,
+						},
+					},
+				},
+			},
+		},
+	}
+	if len(accessRule.Services) == 0 || accessRule.Services[0] == "*" {
+		authzPolicyConfig.Name = fmt.Sprintf("%s-services", namespace)
+		authzPolicyConfig.Spec.(*rbac_v1beta1.AuthorizationPolicy).Selector = nil
+		ug.AuthorizationPolicies = append(ug.AuthorizationPolicies, authzPolicyConfig)
+		return nil
+	}
+	for _, service := range accessRule.Services {
+		serviceName := strings.Split(service, ".")[0]
+		authzPolicyConfig.Name = fmt.Sprintf("%s-%s", serviceName, namespace)
+		workloadSelector, err := ug.getAndAddServiceToWorkloadLabelMapping(service, namespace)
+		if err != nil {
+			return err
+		}
+		authzPolicyCopy, err := copystructure.Copy(&authzPolicyConfig)
+		if err != nil {
+			return err
+		}
+		authzPolicyCopy.(*model.Config).Spec.(*rbac_v1beta1.AuthorizationPolicy).Selector.MatchLabels = workloadSelector
+		authzPolicyConfigCopy := authzPolicyCopy.(*model.Config)
+		ug.AuthorizationPolicies = append(ug.AuthorizationPolicies, *authzPolicyConfigCopy)
+	}
+	return nil
+}
+
+// TODO(pitlv2109): Handle cases with workload selector from destination.labels and other constraints
+// convertAccessRuleToOperation converts one Access Rule to the equivalent Operation.
+func convertAccessRuleToOperation(rule *authz_model.Permission) *rbac_v1beta1.Operation {
+	if rule == nil {
+		return nil
+	}
+	operation := rbac_v1beta1.Operation{}
+	operation.Methods = rule.Methods
+	operation.Paths = rule.Paths
+	// TODO(pitlv2109): Handle destination.port
+	return &operation
+}
+
+// TODO(pitlv2109): Handle properties that are not promoted to first class fields.
+// convertBindingToSources converts Subjects to the equivalent Sources.
+func convertBindingToSources(principals []authz_model.Principal) []*rbac_v1beta1.Rule_From {
+	ruleFrom := []*rbac_v1beta1.Rule_From{}
+	for _, subject := range principals {
+		from := rbac_v1beta1.Rule_From{
+			Source: &rbac_v1beta1.Source{},
+		}
+		if subject.User != "" {
+			from.Source.Principals = []string{subject.User}
+		}
+		if len(subject.Properties) == 0 {
+			ruleFrom = append(ruleFrom, &from)
+			continue
+		}
+		// NOTICE: Only select the first element in the properties list.
+		for k, v := range subject.Properties[0] {
+			switch k {
+			case sourceNamespace:
+				from.Source.Namespaces = v
+			case sourceIP:
+				from.Source.IpBlocks = v
+			case requestAuthPrincipal:
```

Comment:
requestAuthPrincipal should be "iss/sub". I think the current Source.RequestPrincipals field only has "sub"?

Reply:
Hmm, according to the design doc:
`requestPrincipals. A list of request identities (i.e., “iss/sub” claims), which matches “request.auth.principal” attribute`

Reply:
I think you are right.